analysis/furtherModelTesting.ipynb

```diff
@@ -0,0 +1,687 @@
+       "204   4.700000                   100215      1.106129               -5.78  \n",
+       "205   3.000000                      611     41.645220               17.67  \n",
+       "206   1.400000                      300     43.276732               16.79  \n",
+       "\n",
+       "[199 rows x 9 columns]"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "import pandas as pd\n",
+    "import numpy as np\n",
+    "from sklearn.neighbors import KNeighborsRegressor\n",
+    "from sklearn.tree import DecisionTreeRegressor\n",
+    "from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor\n",
+    "from sklearn.metrics import mean_squared_error, mean_absolute_error\n",
+    "from sklearn.model_selection import KFold, train_test_split\n",
+    "\n",
+    "# Drop useless columns\n",
+    "packetTrainingDataPath=\"../training_data/Elysia.Laps.feather\"\n",
+    "df = pd.read_feather(packetTrainingDataPath)\n",
+    "df = df.drop(\n",
+    "        columns=[\n",
+    "            \"msgType\",\n",
+    "            \"_id.$oid\",\n",
+    "            \"averagepackCurrent.$numberDouble\",\n",
+    "            \"timestamp.$numberLong\",\n",
+    "        ]\n",
+    "    )\n",
+    "\n",
+    "#we need the averagepackCurrent data to be numeric instead of {\"$numberDouble\": \"NaN\"}, setting errors='coerce' sets them to numerical NaN\n",
+    "df['averagepackCurrent'] = pd.to_numeric(df['averagepackCurrent'], errors='coerce')\n",
+    "\n",
+    "#drop the 4 rows with null values\n",
+    "df = df.dropna(subset=['distance', 'averagepackCurrent', 'averagespeed'])\n",
+    "\n",
+    "# Remove outliers from the data\n",
+    "#to do this, first define a threshold for outliers (3 standard deviations which contains 99.7% of data)\n",
+    "threshold = 3 * np.std(df['distance'])\n",
+    "\n",
+    "#remove the outliers\n",
+    "df = df[(df['distance'] >= -threshold) & (df['distance'] <= threshold)]\n",
+    "#remove negative distance values\n",
+    "df = df[df['distance'] >= 0]\n",
+    "\n",
+    "#seperate distance from the other features\n",
+    "X = df[['secondsdifference', 'totalpowerin', 'totalpowerout', 'netpowerout', 'amphours', \n",
+    "    'averagepackCurrent', 'batterysecondsremaining', 'averagespeed']]\n",
+    "y = df['distance']\n",
+    "\n",
+    "display(df) #should be 199x9"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## KNN"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 37,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "_knn for KNN is: 0.9679120129289978\n",
+      "MAE for KNN is: 0.5085657314801215\n"
+     ]
+    }
+   ],
+   "source": [
+    "#split into training and test sets\n",
+    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=69420)\n",
```

Comment:
Declare a global variable for random_state=69420 instead of defining it in every single function. Same with test_size

Reply:
Not sure if you've seen or thought about this much, but train_test_split has a `shuffle` argument that is by default `true`. What do you think this does and why do you think it's necessary?